### PR TITLE
benchmark for dvc status with different numbers of ignore rules

### PR DESCRIPTION
**Pull request by karajan1001**
20 ignore rules with 10k files tracked
Screenshot from 2020-05-31 10-16-55

3 ignore rules with 10k files tracked

Screenshot from 2020-05-31 10-17-02

Empty ignore rules file with 10k files tracked

Screenshot from 2020-05-31 10-17-04

### CODE REVIEW

**Comment by karajan1001** [comment]
`dvc add` 10K files takes quite a long time. If I use 40k files, the test would fail because of the asv timeout.

**Comment by skshetry** [comment]
> `dvc add` 10K files takes quite a long time. If I use 40k files, the test would fail because of the asv timeout.

@karajan1001 would it be better to have `add` benchmarks with dvcignore first, rather than status`?

**Comment by karajan1001** [comment]
> > `dvc add` 10K files takes quite a long time. If I use 40k files, the test would fail because of the asv timeout.
> 
> @karajan1001 would it be better to have `add` benchmarks with dvcignore first, rather than status`?

From my understanding,  `.dvcignore` influences both `dvc add` and `dvc status`. But the `dvc add` takes much longer time and `.dvcignore` is not the bottleneck it. It is better to test `.dvcignore` with `dvc status` rather than `dvc add`. 

But maybe I could `dvc add data` first and then copy the files.

**Comment by karajan1001** [comment]
> `dvc add` 10K files takes quite a long time. If I use 40k files, the test would fail because of the asv timeout.

This is because the test runs for several times. And in each time calculating md5 and adding files to the cache took server seconds. Now I put `dvc add` before copying files to data, this change can prevent adding files to the cache.

**Comment by karajan1001** [comment]
> > `dvc add` 10K files takes quite a long time. If I use 40k files, the test would fail because of the asv timeout.
> 
> This is because the test runs for several times. And in each time calculating md5 and adding files to the cache took server seconds. Now I put `dvc add` before copying files to data, this change can prevent adding files to the cache.

After these changes, here is the result. 


20 ignore rules with 10k files tracked


Screenshot from 2020-05-31 14-19-15
3 ignore rules with 10k files tracked

Screenshot from 2020-05-31 14-19-18

Empty ignore rules file with 10k files tracked

Screenshot from 2020-05-31 14-19-19

These results matches 
Screenshot from 2020-05-31 14-27-27

**Comment by karajan1001** [comment]
Add one more test 1 ignore file with 30 rules compare to 3 ignore file with 10 rules each.

1 file 30 rules
Screenshot from 2020-05-31 15-54-41

3 file each has 10 rules
Screenshot from 2020-05-31 15-54-43

File: benchmarks/status.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    def setup(self):
+        super().setup()
+        data_path = os.path.join(self.test_directory.name, "data")
+        data_data_path = os.path.join(data_path, "data")
+        self.add_ignore_rules(self.test_directory.name, 10)
+        self.add_ignore_rules(data_path, 10)
+        self.add_ignore_rules(data_data_path, 10)
+
+    def time_status(self):
+        assert main(["status", "--quiet"]) == 1
```
Comment (pared) [inline benchmarks/status.py:59]: Do we need this one? 
While I believe benchmark is a necessity for `dvcignore` I am not sure this one is a required test if `DvcIgnore30Rules` already exists.

 We are using `dvc-bench` to observe the influence of implemented changes on different scenarios behavior. I believe that having `DvcIgnore3x10` and `DvcIgnore30` will always show us the difference between iterating over one list vs iterating over 3 lists. 
The underlying implementation of `DvcIgnorePatterns` will be the same for both cases, so even if we introduce some change that will slow it down, the same degradation will be caught by both benchmarks.

Comment (karajan1001) [reply benchmarks/status.py:59]: > We are using `dvc-bench` to observe the influence of implemented changes on different scenarios behavior. I believe that having `DvcIgnore3x10` and `DvcIgnore30` will always show us the difference between iterating over one list vs iterating over 3 lists.

@pared 
You are right. Although, it shows a big performance gap between `DvcIgnore3x10` and `DvcIgnore30`.But, after we solved this problem, they would always show synchronized performance.

**Comment by karajan1001** [comment]
Please wait for a moment.
This benchmark runs too long and itself needs optimization.
1.  Use a module-level setup and teardown. 
2.  Set `repeat` to 1 and increase `numbers` (1 `setup` and run tests multi-times) [ASV](https://asv.readthedocs.io/en/stable/benchmarks.html?highlight=repeat%20number#benchmark-attributes)

**Comment by karajan1001** [comment]
> Please wait for a moment.
> This benchmark runs too long and itself needs optimization.
> 
> 1. Use a module-level setup and teardown.
> 2. Set `repeat` to 1 and increase `numbers` (1 `setup` and run tests multi-times) [ASV](https://asv.readthedocs.io/en/stable/benchmarks.html?highlight=repeat%20number#benchmark-attributes)

1. module-level setup and teardown runs in every benchmark. They are combined with class-level ones. 
2. I had set repeat to 1 and numbers to 10, and the total running time cut in half on my computer. But it seems there is still something wrong with it. I had submitted an issue to [asv]( https://github.com/airspeed-velocity/asv/issues/930)

**Comment by pared** [comment]
@karajan1001 
Asv has something called `setup_cache`, some context [here](https://asv.readthedocs.io/en/stable/writing_benchmarks.html#setup-and-teardown-functions).

The problem with it is that this method writes pickled setup result to disk, and loads it upon next execution of benchmark. In case of DVC, that does not make sense as we are operating on a directory/repository which changes state during benchmark execution.

[Some more context](https://github.com/iterative/dvc-bench/issues/12)

**Comment by karajan1001** [comment]
> @karajan1001
> Asv has something called `setup_cache`, some context [here](https://asv.readthedocs.io/en/stable/writing_benchmarks.html#setup-and-teardown-functions).
> 
> The problem with it is that this method writes pickled setup result to disk, and loads it upon next execution of benchmark. In case of DVC, that does not make sense as we are operating on a directory/repository which changes state during benchmark execution.
> 
> [Some more context](https://github.com/iterative/dvc-bench/issues/12)

Yes,  our setup is of paths and files, not python in-memory objects. 

Currently, the whole bench takes 10 minutes on my computer, only a quarter compared to the first edition.

There is only one problem remains that each `setup()` runs twice.

**Comment by pared** [comment]
@karajan1001 
It's most likely due to fact that we run not only benchmarks. If you inspect `run_benchmarks.dvc` you will note that we also profile the tests. That is the reason for the additional run.

**Comment by karajan1001** [comment]
> @karajan1001
> It's most likely due to fact that we run not only benchmarks. If you inspect `run_benchmarks.dvc` you will note that we also profile the tests. That is the reason for the additional run.

Thank you @pared you gave me the right answer. According to [ASV](https://asv.readthedocs.io/en/stable/commands.html?highlight=profile#asv-run)
```
  --profile, -p         In addition to timing, run the benchmarks through the
                        `cProfile` profiler and store the results.
```
And I tested that in deleting `--profile` the second `setup()` disappeared. I have nothing to optimize now.

**Comment by pared** [comment]
@karajan1001 also, I created #33, not sure whether this profiling makes sense, maybe we should do it only when we notice degradation in performance.

File: benchmarks/status.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    def setup(self):
+        super().setup()
+        self.repo = init_dvc(self.test_directory.name)
+        dataset_path = random_data_dir(10000, 10)
+        os.makedirs(
+            os.path.join(self.test_directory.name, "data"), exist_ok=True
+        )
+        assert main(["add", "data", "--quiet"]) == 0
+        shutil.copytree(dataset_path, "data/data")
+        # calculating md5
```
Comment (pared) [inline benchmarks/status.py:28]: +1

**Comment by pared** [comment]
Now that I think about it, shouldn't we just have a single benchmark, with, multiple ignore rules?

I mean having 0, 3 and 30 seems to me like creating tests to compare `status` performance between those 3 use cases. 

The point of `dvc-bench` is to detect degradation in performance (and not see how `dvcignore` influences walk through directory), and for that we need benchmark checking status with any existing rules. And having 0 `dvcignore` rules looks to me like a status benchmark (which we don't currently have). 

I would consider leaving 30 rules case as is, removing 3 rules use case and moving 0 cases to `StatusBenchmark` (not necessarily in this change).

What do you think @karajan1001?

**Comment by karajan1001** [comment]
> @karajan1001 also, I created #33, not sure whether this profiling makes sense, maybe we should do it only when we notice degradation in performance.

Sorry for the late reply, I agreed with this. 


> Now that I think about it, shouldn't we just have a single benchmark, with, multiple ignore rules?
> 
> I mean having 0, 3 and 30 seems to me like creating tests to compare `status` performance between those 3 use cases.
> 
> The point of `dvc-bench` is to detect degradation in performance (and not see how `dvcignore` influences walk through directory), and for that we need benchmark checking status with any existing rules. And having 0 `dvcignore` rules looks to me like a status benchmark (which we don't currently have).
> 
> I would consider leaving 30 rules case as is, removing 3 rules use case and moving 0 cases to `StatusBenchmark` (not necessarily in this change).
> 
> What do you think @karajan1001?

Yes, the zero  `dvcignore` is more a status benchmark.  I had changed the `DVCIgnoreEmpty` name several times and couldn't make 
a final decision. 

30 rules comparing 3 rules can tell us how the time cost grows with the increase of `dvcignore`.  I think it is OK if the time cost grows linear to the number of rules in `dvcignore`. Without this assumption, we could not separate the influence of preprocessing and regex matching. But maybe this detailed performance influences should only be checked if there is an obvious performance degradation. Just like the `--profile` in the previous discussion. Benchmark gives a big picture, only some degradation detected should we go into the small pieces.

**Comment by pared** [comment]
@karajan1001 
>But maybe this detailed performance influences should only be checked if there is an obvious performance degradation. Just like the --profile in the previous discussion. Benchmark gives a big picture, only some degradation detected should we go into the small pieces.

Please take a note of @efiop comment under the profile issue.  We can leave the profiles, in order to have analysis materials if something goes south.

**Comment by karajan1001** [comment]
> > But maybe this detailed performance influences should only be checked if there is an obvious performance degradation. Just like the --profile in the previous discussion. Benchmark gives a big picture, only some degradation detected should we go into the small pieces.

@pared @efiop 
And what about the 3 ignore rules benchmark? In my opinion with the message from `--profile`, we already have enough information.

**Comment by pared** [comment]
@karajan1001 

I agree, to see the detailed difference in case 3 vs 30, we would need to compare charts for both tests. And the main point of this repo is to fulfill #8, which will be hard to programmatically define in case of comparing 2 benchmarks. 

So let's get rid of it for now?

**Comment by pared** [comment]
@karajan1001 great change! Thank you very much!

**Pull request merged**